ephios/plugins/questionnaires/models.py

```diff
@@ -0,0 +1,162 @@
+
+    class Meta:
+        verbose_name = _("Question")
+        verbose_name_plural = _("Questions")
+
+    def __str__(self):
+        return str(self.name)
+
+    def get_signup_form_field(
+        self, participant: AbstractParticipant, participation: AbstractParticipation, signup_choice
+    ):
+        # Restore answer from participation (when editing) or the user's saved answers (if local participant)
+        if existing_answer := Answer.objects.filter(
+            participation_id=participation.pk, question=self
+        ).first():
+            initial = existing_answer.answer
+        elif saved_answer := (
+            SavedAnswer.objects.filter(user=participant.user, question=self).first()
+            if isinstance(participant, LocalUserParticipant)
+            else None
+        ):
+            initial = saved_answer.answer
+        else:
+            initial = None
+
+        # Reset answer if selected option does not exist
+        if self.type in [self.Type.SINGLE, self.Type.MULTIPLE] and initial not in self.choices:
+            initial = None
+
+        field_name = self.get_form_slug()
+        required = self.required if initial is None else False
+
+        field = {
+            "form_class": None,
+            "form_kwargs": {},
+            "serializer_class": None,
+            "serializer_kwargs": {},
+        }
+
+        field["form_kwargs"] = {
+            "label": self.question_text,
+            "help_text": self.description,
+            "initial": initial,
+            "required": required,
+        }
+
+        field["serializer_kwargs"] = {
+            "required": required,
+        }
+
+        match self.type:
+            case self.Type.TEXT:
+                field["form_class"] = forms.CharField
+                field["serializer_class"] = serializers.CharField
+                # pylint: disable=protected-access
+                max_length = Answer._meta.get_field("answer").max_length
+                field["form_kwargs"]["max_length"] = max_length
+                field["serializer_kwargs"]["max_length"] = max_length
+            case self.Type.SINGLE:
+                field["form_class"] = forms.ChoiceField
+                field["serializer_class"] = serializers.ChoiceField
+                field["form_kwargs"]["widget"] = (
+                    forms.RadioSelect if len(self.choices) <= 5 else Select2Widget
+                )
+            case self.Type.MULTIPLE:
+                field["form_class"] = forms.MultipleChoiceField
+                field["serializer_class"] = serializers.MultipleChoiceField
+                field["form_kwargs"]["widget"] = Select2MultipleWidget
+
+        if self.type != self.Type.TEXT:
+            assert isinstance(
+                self.choices, list
+            ), f"The choices of question {self.name} are not a list"
+            # pylint: disable=not-an-iterable
+            choices = [(choice, choice) for choice in self.choices]
```

Comment:
this looks dangerous, can every arbitrary string be a choice key in HTML selects? What if it contains special characters?

Reply:
Added comment in 15619e09bb9d5df943f24692cd0c3de89bfde34f